tools/linter/lint.py

```diff
@@ -0,0 +1,258 @@
+        return f"{self.name} using {self.exe}"
+
+
+def build_parser(linters) -> argparse.Namespace:
+    parser = argparse.ArgumentParser("Run linters on PyTorch", add_help=False)
+
+    # Setting default values here will prevent linters from overriding option values
+    # For example, it doesn't make sense for a linter to override
+    # --all set to true because this breaks the semantics of the
+    # linter running on all filtered files by default
+    parser.add_argument(
+        "--all",
+        help="Run linters without filtering for changed files",
+        default=False,
+        action="store_true",
+    )
+    parser.add_argument(
+        "-g",
+        "--glob",
+        action=Glob2RegexAction,
+        help="Only lint files that match these glob patterns "
+        "(see documentation for `fnmatch` for supported syntax)."
+        "If a pattern starts with a - the search is negated for that pattern.",
+    )
+    parser.add_argument(
+        "-x",
+        "--regex",
+        action="append",
+        help="Only lint files that match these regular expressions (from the start of the filename). "
+        "If a pattern starts with a - the search is negated for that pattern.",
+    )
+    parser.add_argument(
+        "-p", "--paths", nargs="+", help="Lint only the given paths (recursively)",
+    )
+    parser.add_argument(
+        "-d",
+        "--diff-file",
+        default=None,
+        help="Lint files filtered by a file containing a diff",
+    )
+    parser.add_argument(
+        "-v",
+        "--verbose",
+        action="store_true",
+        help="Print verbose output",
+        default=False,
+    )
+
+    subparsers = parser.add_subparsers()
+    for linter in linters:
+        subparser = subparsers.add_parser(
+            linter.name, parents=[parser], help=f"Run {linter.name} on the codebase"
+        )
+        linter.build_parser(subparser)
+        subparser.set_defaults(linter=linter)
+
+    return parser
+
+
+def init_linter(name: str) -> Linter:
+    module_name = kebab2snake(name)
+    lint_class_name = kebab2camel(name)
+
+    module = importlib.import_module(module_name)
+    Linter = module.__dict__[lint_class_name]
+
+    return Linter()
+
+
+def init_linters() -> List[Linter]:
+    return [init_linter(name) for name in SUPPORTED_LINTERS]
+
+
+def merge_namespaces(a, b):
+    a_dict = vars(a)
+    b_dict = vars(b)
+    c = a_dict.copy()
+    for (k, v) in b_dict.items():
+        if v is not None or k not in c:
+            c[k] = v
+    return argparse.Namespace(**c)
+
+
+async def main() -> None:
+    linters = init_linters()
+    parser = build_parser(linters)
+    options = parser.parse_args()
+
+    if not options.all:
+        options.paths = await find_changed_files()
+
+    # Run linter
+    if "linter" in options:
+        linters = [options.linter]
+
+    result = CommandResult(0, "", "")
+
+    # We run the linters sequentially because the logging infrastructure for
```

Comment:
What do you think about the approach of `actions_local_runner.py` where it saves all the output in memory and then writes it all at once inside an `async` block (so not truly parallel) to avoid interleaving?